CPS-0003? | Smart tokens

rphair opened the pull request:
Cloned from commit history of defunct branch of @szist's original (abandoned) https://github.com/cardano-foundation/CIPs/pull/382.

We decided at the CIP meeting today to resuscitate this to support cooperative work on these candidate CIPs:
- https://github.com/cardano-foundation/CIPs/pull/444
- https://github.com/cardano-foundation/CIPs/pull/832
- https://github.com/cardano-foundation/CIPs/pull/944

@colll78 @Crypto2099 have offered to check technical & ecosystem content to ensure it remains valid 2+ years after the original submission.  I've cleaned up some missing items and removed all TBDs, which further review is welcome to continue:
- _Use cases_ > "add more examples"
- _Identifying smart tokens_ > "add more considerations"
- _Open Questions_ > ...

---

([rendered from this PR's branch](https://github.com/rphair/CIPs/blob/cps-0003-reconstruction/CPS-0003/README.md))

rphair commented:
thanks @fallen-icarus - I was just reading your https://github.com/cardano-foundation/CIPs/pull/944#discussion_r1880802631 and noting this question to keep discussing for all CIPs related to this one.  I've added it to the _Open Questions_ section here: https://github.com/cardano-foundation/CIPs/pull/947/commits/97d90b87609d7f475e204a7fc10ce3df69d479ff

Ryun1 commented on CPS-0003/README.md:
````diff
@@ -0,0 +1,141 @@
+**2. add policy headers**
+This would be a breaking change to the structures and policies. Basically it would be redefining the multiasset structure to:
+
+```
````
```suggestion
```cddl
```

rphair replied:
great; I don't think we had the `cddl` GitHub language type 2 years ago 😅

Ryun1 commented on CPS-0003/README.md:
````diff
@@ -0,0 +1,141 @@
+
+In this situation, similar to how value can be just a coin or an array, the `policy_id` could be redefiend to be:
+
+```
````
```suggestion
```cddl
```

rphair commented:
CIP meeting yesterday emphasised that we are waiting for @Crypto2099's broad & historical perspective to see if any updates need to be made to the background material, motivation, etc. based on the 2 years that have passed since original proposal, as per https://github.com/cardano-foundation/CIPs/pull/947#issue-2730776457 - as soon as that's cleared we can merge this, likely at the next CIP meeting.  Thanks @colll78 for providing your own confirmation.

Crypto2099 commented on CPS-0003/README.md:
````diff
@@ -0,0 +1,141 @@
+value = coin / [coin,multiasset<uint>] / [coin, multiasset<uint>, smartasset<uint>]
+mint = multiasset<int64> / [multiasset<uint>, smartasset<uint>]
+```
+
+Alternatively the value could be "objectified" similar to babbage tx outputs:
+
+```cddl
+value = {
+  0: coin,
+  ? 1: multiasset<uint>,
+  ? 2: smartasset<uint>
+}
+```
+
+This explicit segregation would have the benefit that it could be easier backwards compatible. In this format also the policy ids could remain as `hash28` script hashes.
+
+**2. add policy headers**
+This would be a breaking change to the structures and policies. Basically it would be redefining the multiasset structure to:
+
+```cddl
+multiasset<a> = { * policy_id_with_header => { * asset_name => a } }
+; token identifier format:
+; [ 8 bit header | policy_id ];
+; 0 - open transfer token
+; 1 - smart token
+policy_id_with_header = bytes
+policy_id = scripthash
+```
+
+**3. policy id length**
+
+In this situation, similar to how value can be just a coin or an array, the `policy_id` could be redefiend to be:
+
+```cddl
+policy_id = scripthash / bytes ; as above [ 8 bit header | scripthash ]
+```
+
+Both 2. and 3. are relatively error-prone, as many projects and tools assume 28byte policy IDs. Especially when encoding them as `subject` - for example for the offchain registry, there would be no way to tell, if the policyID is with the header or not.
+
+**TBD** add more considerations
````
I believe this section can be removed as the author has not actively maintained it and it should be its own CIP that seeks to address one or more of the open questions listed below.

rphair replied:
@Crypto2099 then let's not merge this until we decide what to do with the omitted material, to ensure we don't lose track of it... and let's talk about that at the meeting coming up this hour.

rphair replied:
Addressed with https://github.com/cardano-foundation/CIPs/pull/947/commits/5d2369cfe0385d7483478137418426b805110ebb as documented in https://github.com/cardano-foundation/CIPs/issues/969.

Crypto2099 commented on CPS-0003/README.md:
```diff
@@ -0,0 +1,141 @@
+Authors:
+    - Istvan Szentandrasi <szist@zoeldev.com>
+Proposed Solutions: []
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/382
```
There are certainly several other "conversations" around smart tokens at this point as I believe we have no less than 3 numbered CIPs that seek to address this behavior. Should they be included here as part of the discussion or should they be referenced somewhere in the body of the document itself?

rphair replied:
@Crypto2099 since the CPS at some point will include all those as CIP "mapping" links, I think the body content should enumerate the 3 relevant CIPs which in turn will have their own PR discussion links... and will have others added as time goes on, perhaps.  If & when each of those CIP PRs is merged, they can be replaced by CIP document links as we've done in other cases.  Once we can target a section to add these (at the meeting in the next hour) I'll update the content as such.

rphair replied:
Addressed with https://github.com/cardano-foundation/CIPs/pull/947/commits/565e374d65805381cce2bff27eaae1b5e2a989a1

rphair commented:
@Crypto2099 @Ryun1 I've cleaned up the 2 outstanding issues from today's CIP meeting.

I also added myself as a less significant co-author because otherwise the unique "posthumous" activity of this document would go without a visible hand in reconstituting all this material & ensuring that future updates will make sense until it's obvious from straightforward GitHub history.  Anyone who feels this is inappropriate please indicate so & I'll remove it (cc @szist).

rphair commented:
p.s. @Ryun1 @Crypto2099 @perturbing since this is the first time a candidate CIP or CPS PR has been added to the YAML front matter, I need to point out the changes in https://github.com/cardano-foundation/CIPs/pull/947/commits/565e374d65805381cce2bff27eaae1b5e2a989a1 & hold up for debate:

Although it would make more sense to put `?` after the CIP name; e.g.:
```
Proposed Solutions:
    - CIP-0113?: https://github.com/cardano-foundation/CIPs/pull/444
    - CIP-0125?: https://github.com/cardano-foundation/CIPs/pull/832
    - CIP-0143?: https://github.com/cardano-foundation/CIPs/pull/944
```
... although this would properly & rigorously emphasise that these documents _aren't CIPs yet_:
- without double quotes I think it's a YAML syntax error;
- the (perhaps) required double quotes around everything before the `:` looks awkward & ugly;
- even if it's not a YAML error the punctuation still "looks ugly" and will leave lots of readers not understanding what it means. 😅
